### PR TITLE
Fix dispatch bugs

### PR DESCRIPTION
**Pull request by pengzhout**

### CODE REVIEW

**Comment by pengzhout** [comment]
@xiong-gang @eurekaka @yydzero @hlinnaka 
Please help to review this if possible,  thanks.

File: src/backend/cdb/dispatcher/cdbgang_async.c [hunk 1]
```diff
@@ -56,212 +58,199 @@ createGang_async(GangType type, int gang_id, int size, int content)
...
-			 * valid segdb we error out.  Also, if this segdb is invalid, we must
-			 * fail the connection.
-			 */
-			segdbDesc = &newGangDefinition->db_descriptors[i];
+		/*
+		 * Build the connection string.  Writer-ness needs to be processed
+		 * early enough now some locks are taken before command line options
+		 * are recognized.
+		 */
+		build_gpqeid_param(gpqeid, sizeof(gpqeid),
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbgang_async.c]: I think it would make sense to push this build_gpqeid_param() call into cdbconn_doConnectStart() itself. It's used by all callers of cdbconn_doConnect() and cdbconn_doConnectStart(), so it seems unnecessary for all the callers to do the same dance.

(not directly related to this patch, of course, just something that caught my eye)

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 2]
```diff
@@ -872,81 +872,6 @@ static void addGangToAllocated(Gang *gp)
...
- * as the dispatch agent doesn't differentiate allocated from available, or 1 gangs
- * from n-gangs.  It assumes the QD keeps track of all that.
- *
- * It might be nice to pass gang type and size to this routine as safety checks.
- *
- * TODO: dispatch agent. remove or re-implement.
- */
-
-Gang *
-findGangById(int gang_id)
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbgang.c]: This is the same change as PR #1016.  +1 on doing this. Can you push PR #1016 first, please, so that we can get it off this PR. This is large enough without it :-).

File: src/backend/postmaster/postmaster.c [hunk 3]
```diff
@@ -4219,6 +4219,7 @@ do_immediate_shutdown_reaper(void)
 
         zeroIfPidEqual(pid, &BgWriterPID);
         zeroIfPidEqual(pid, &CheckpointPID);
+		zeroIfPidEqual(pid, &WalWriterPID);
```
Comment (hlinnaka) [inline src/backend/postmaster/postmaster.c]: These are the same changes as in PR #1039 and PR #906. +1 on doing this, but seems unrelated to this PR.

File: src/backend/cdb/dispatcher/cdbgang_thread.c [hunk 4]
```diff
@@ -189,55 +187,54 @@ createGang_thread(GangType type, int gang_id, int size, int content)
...
-		newGangDefinition = NULL;
+		if (successful_connections + in_recovery_mode_count == size)
+		{	
+			/*
+			 * Retry for non-writer gangs is meaningless because
+			 * writer gang must has gone when QE is in recovery mode
+			 */
+			if (type != GANGTYPE_PRIMARY_WRITER)
+			{
+				appendPQExpBuffer(&create_gang_error, "segments are in recovery mode, writer gang lost\n");
```
Comment (hlinnaka) [inline src/backend/cdb/dispatcher/cdbgang_thread.c]: Does "recovery mode" mean WAL recovery, i.e. "InRecovery == true", here? Or something else?

**Comment by hlinnaka** [comment]
Lots of little changes. I don't really understand all the bugs this is fixing, but I don't see anything wrong either. So I guess that's a +1 from me.

File: src/backend/postmaster/postmaster.c (hunk 3 above)
Comment (pengzhout) [inline src/backend/postmaster/postmaster.c]: Sure, i will not submit this commit, just temporary make postmaster rest works,  otherwise all my test cases will fail

File: src/backend/cdb/dispatcher/cdbgang_thread.c (hunk 4 above)
Comment (pengzhout) [inline src/backend/cdb/dispatcher/cdbgang_thread.c]: i thinks it means segment is under crash recovery which should include WAL recovery and other steps like backends shutting down, auxiliary backend restart, shared memory reset etc. it comes from the message POSTMASTER_IN_RECOVERY_MSG.

File: src/backend/cdb/dispatcher/cdbgang_async.c (hunk 1 above)
Comment (pengzhout) [inline src/backend/cdb/dispatcher/cdbgang_async.c]: thanks, worth a patch for this

File: src/backend/cdb/dispatcher/cdbgang_thread.c [hunk 5]
```diff
@@ -189,55 +187,54 @@ createGang_thread(GangType type, int gang_id, int size, int content)
...
+				goto exit;	
+			}
 
-		ELOG_DISPATCHER_DEBUG("createGang: gang creation failed, but retryable.");
+			ereport(WARNING, (errmsg("segments are in recovery mode, sleep %d microseconds and retry %d of %d ",
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbgang_thread.c]: it would be too many warning messages to the front end. I'd suggest LOG.

**Comment by pengzhout** [comment]
merged and close, thanks for review
